src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs

```diff
@@ -173,65 +174,75 @@ public static async Task WriteResponseMessage(HttpResponse response, Encoding en
             IMessage requestMessage;
             if (serverCallContext.DescriptorInfo.BodyDescriptor != null)
             {
-                if (!serverCallContext.IsJsonRequestContent)
+                Type type;
+                object bodyContent;
+
+                if (serverCallContext.DescriptorInfo.BodyDescriptor.FullName == HttpBody.Descriptor.FullName)
                 {
-                    GrpcServerLog.UnsupportedRequestContentType(serverCallContext.Logger, serverCallContext.HttpContext.Request.ContentType);
-                    throw new RpcException(new Status(StatusCode.InvalidArgument, "Request content-type of application/json is required."));
+                    type = typeof(HttpBody);
+
+                    bodyContent = await ReadHttpBodyAsync(serverCallContext);
                 }
+                else
+                {
+                    if (!serverCallContext.IsJsonRequestContent)
+                    {
+                        GrpcServerLog.UnsupportedRequestContentType(serverCallContext.Logger, serverCallContext.HttpContext.Request.ContentType);
+                        throw new InvalidOperationException("Request content-type of application/json is required.");
```

Comment:
Why did this change to an `InvalidOperationException`? It seems like an odd choice for invalid input from the client. Is that what this is?

It looks like it's still getting wrapped in an `RpcException`, but it's still an odd `InnerException`.

Reply:
What do you suggest instead?

Reply:
I just noticed it looks like this isn't even an `InnerException`, but instead it's just used for to copy its `Message` to the `RpcException` that's ultimately thrown. Now I care less than before.

It's kinda gross the flow the exception message this way to the common throwing logic, but if that's really the best way to do it, just leave it as `Exception` instead of some derived type that makes no difference.

Reply:
Exception is now preserved by setting it as DebugException on RpcException.Status. It's a little weird, but the best I can do.

I looked at the JSON helper methods for serializing and deserializing from a HttpContext (SerializeAsJson, DeserializeFromJson) and they use InvalidOperationException.